tests/integration/azure/test_azure_credentials.py

```diff
@@ -0,0 +1,142 @@
+import base64
+import re
+from unittest.mock import MagicMock
+
+import pytest
+from databricks.labs.blueprint.tui import MockPrompts
+
+from databricks.labs.ucx.assessment.azure import AzureServicePrincipalInfo
+from databricks.labs.ucx.assessment.crawlers import _SECRET_PATTERN
+from databricks.labs.ucx.azure.access import StoragePermissionMapping
+from databricks.labs.ucx.azure.azure_credentials import AzureServicePrincipalMigration
+
+
+@pytest.fixture
+def prepare_spn_migration_test(ws, debug_env, make_random):
+    def inner(read_only=False) -> dict:
+        spark_conf = ws.clusters.get(cluster_id=debug_env["TEST_LEGACY_SPN_CLUSTER_ID"]).spark_conf
+
+        application_id = spark_conf.get("fs.azure.account.oauth2.client.id")
+
+        secret_matched = re.search(_SECRET_PATTERN, spark_conf.get("fs.azure.account.oauth2.client.secret"))
+        if secret_matched:
+            secret_scope, secret_key = (
+                secret_matched.group(1).split("/")[1],
+                secret_matched.group(1).split("/")[2],
+            )
+        assert secret_scope is not None
+        assert secret_key is not None
+
+        secret_response = ws.secrets.get_secret(secret_scope, secret_key)
+        client_secret = base64.b64decode(secret_response.value).decode("utf-8")
+
+        end_point = spark_conf.get("fs.azure.account.oauth2.client.endpoint")
+        directory_id = end_point.split("/")[3]
+
+        name = f"testinfra_storageaccess_{make_random(4).lower()}"
+
+        azure_resource_permissions = MagicMock()
+        azure_resource_permissions.load.return_value = [
+            StoragePermissionMapping(
+                prefix="abfss://things@labsazurethings.dfs.core.windows.net/avoid_ext_loc_overlap",
+                client_id=application_id,
+                principal=name,
+                privilege="READ_FILES" if read_only else "WRITE_FILES",
+                directory_id=directory_id,
+            )
+        ]
+
+        azure_sp_crawler = MagicMock()
+        azure_sp_crawler.snapshot.return_value = [
+            AzureServicePrincipalInfo(
+                application_id=application_id,
+                secret_scope=secret_scope,
+                secret_key=secret_key,
+                tenant_id="test",
+                storage_account="test",
+            )
+        ]
+
+        installation = MagicMock()
+        installation.save.return_value = "azure_service_principal_migration_result.csv"
+
+        return {
```

Comment:
Don't return it a bunch of variables as untyped dictionary. Refactor to return one or two variables. 

This is impossible to maintain

Reply:
fixed